enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,169 @@
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of OpenShift 4 is to make upgrades seamless and risk free.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give the operators ability to run some checks before the upgrade to find out if the new incoming version is compatible with the existing version.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before going through the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered check if upgrade to the target version is going to be successfully without doing an upgrade.
+Use preflight checks to improve the success rate of upgrades.
+
+
+### Goals
+
+* This enhancement proposes a framework for preflight checks and discusses the CVO side implementation of that framework.
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+* Downloading and verification of images associated with release payload should be part of preflight check.
+
+### Non-Goals
+
+* This enhancement does not commit particular operators for providing preflight implementations or discuss the operator-specific logic that should be executed by preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+* It is not the responsibility of preflight checks to protect against upgrades involving more than one minor version i.e. 4.y.z to 4.(y+2).z.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The bin path need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+/bin/preflight must be executable. It can be symbolic link to the operator binary or a script or It can be non existent.
+Users can skip or override the preflight checks with `--force`.
+
+### User Stories
+
+#### Story 1
+
+Some customers would prefer to disable automatic credential provisioning by Cloud credential operator (CCO) and instead manage it themselves.
+However if new permissions are required by the target version, the upgrade could block midway even if the administrator provisioned more powerful credentials.
+With the help of preflight checks CCO can check if they can safely upgrade to a given release before doing the actual upgrade.
```

Comment:
Why is the operator not reporting upgradable false?

Reply:
The pre-flights supposed to tell the customer what will be set to ` upgradable false` if upgrade  actually started. Now customer has information without doing the actual upgrade. Now customer can fix the cluster and start the upgrade. Also this check can be done against multiple version to see which versions can be upgraded without any issue. Preflight checks will not set upgradable false` for both z stream or y stream.

Reply:
I think the cred folks actually [are][1] moving to `Upgradeable=False` for this use case.  The downside is that they need to backport knowledge of incoming 4.(y+1) credential requests to the 4.y.z release, and that means raising the minimum 4.y.z from which you can update to 4.(y+1).  But that's not a terrible problem.

[1]: https://bugzilla.redhat.com/show_bug.cgi?id=1871713#c7

Reply:
Yes it only recently came to our attention that upgradable=false does NOT impact .z upgrades and would only block y upgrades. This largely solves our use case, it does require us to push code to stable streams to check for what is coming in 4.6, and it puts that burden on our team to monitor component cred changes and implement the backwards checks when we a change and get it pushed to a .z, as opposed to a generic solution we could implement with preflights by examining the upcoming release image. We can live with this and see how it goes.

Reply:
So are we going to remove this use case or put it in “solved by upgradeable”?